File: leetcodes/207. Course Schedule/courseSchedule.java
import java.util.Queue;

class Solution {
  public boolean canFinish(int numCourses, int[][] prerequisites) {
    //BFS : time O(V+E) space: O(V)
    /**
     * Queue = new LinkedList<>() -> start from the must take vertice for multiple courses
     * List<Int>[] MustTake = new LinkedList[numCourses] (vertice, adj vertice:what is want take)
     * int[] Indegree = new int[numCourses]; (vertice, edge: how many must take)
     * queue.offer(i)-> when indegree[i] = 0 =>no incoming edge
     * while queue(mustTake)
     *  course = poll queue, count++
     *  for each MustTake[course] element -> if -- Indegree[elem] = 0 ->  queue.offer(elem)
     *  ex: n =3, [[0,1],[0,2]] ->musttake[1] = [0]; musttake[2] = [0] , indegree[0] = 2
     *  ex: n = 3 [[0,1],[2,1]] ->musttake[1] = [0,2]
     */
    List<Integer>[] graph = new LinkedList[numCourses];
    Queue<Integer> queue = new LinkedList<>();
    int[] indegree = new int[numCourses];
    int count = 0;
    for (int i = 0; i < numCourses; i++) graph[i] = new LinkedList<>();
    for (int[] pair : prerequisites) {
      graph[pair[1]].add(pair[0]);
      indegree[pair[0]]++;
    }
    for (int i = 0; i < indegree.length; i++)
      if (indegree[i] == 0) queue.offer(i);
    while (!queue.isEmpty()){
      int course = queue.poll();
      count ++;
      for (int wannaTake : graph[course])
        if (--indegree[wannaTake] == 0) queue.offer(wannaTake);
    }
      return count == numCourses;
  }
}

File: leetcodes/207. Course Schedule/courseSchedule.py
import collections
from typing import Deque


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
      adjMap =  [[] for _ in range(numCourses)]
      indegree = [0] * numCourses
      for c1, c2 in prerequisites:
        adjMap[c2].append(c1)
        indegree[c1] += 1

      queue = collections.deque(i for (i, d) in enumerate(indegree) if d == 0)
      count = 0

      while queue:
        cur = queue.popleft()
        count += 1
        for wannaTake in adjMap[cur]:
          indegree[wannaTake] -= 1
          if indegree[wannaTake] == 0: queue.append(wannaTake) #don't queue wannaTake dupliacte

      return count == numCourses

      '''
        was using dict() but kinda complex to loop thru
        and then use adjMap array based on index to store array of courses
        need to consider edges vertice relationsip
        indegree is to store the incoming edge for each vertex if there is 0 -> possible no cycle loop
           then we are able to get the wannaTake courses to count the total courses
        if incoming edge is not zero -> there must be a cycle loop =>false
        time O(v+ E(last loop?)) space O(v)
      '''